### PR TITLE
Added failure enum options to DerivativeParsedMaterial

### LINKED ISSUE #14975: Allow exceptions instead of errors for fparser evalutations
## Reason
There are times where an exception would help continue simulations if an error occurs during function parsing, e.g. log of a negative value or nan. Specifically, change the behavior of `FunctionParserUtils` for `DerivativeParsedMaterials`.

## Design
Option 1) Either make a new bool in `FunctionParserUtils` to indicate throwing an exception rather than an error.

Option 2) Deprecate `fail_on_evalerror`, and create a new enum `evalerror_action`, or something like that, that has `nothing warning error exception` with default `nothing`, similar to `fail_on_evalerror`

ping @dschwen for decision between option 1 and 2.

## Impact
Flexibility in `DerivativeParsedMaterials`

### PR DESCRIPTION
**Pull request by tophmatthews**
Closes #14975

### CODE REVIEW

**Comment by tophmatthews** [comment]
ping @dschwen

File: framework/include/utils/FunctionParserUtils.h [hunk 1]
```diff
@@ -44,7 +44,7 @@ class FunctionParserUtils
 
 protected:
   /// Evaluate FParser object and check EvalError
-  Real evaluate(ADFunctionPtr &);
+  Real evaluate(ADFunctionPtr &, std::string object_name = "");
```
Comment (dschwen) [inline framework/include/utils/FunctionParserUtils.h]: Let's not pass strings by value

Comment (tophmatthews) [reply framework/include/utils/FunctionParserUtils.h]: `const std::string & object_name = ""` should do it right?

**Review by dschwen — APPROVED** [review]
This is great!

**Comment by tophmatthews** [comment]
Should there be a date somewhere for when to take out the deprecated params?

**Comment by dschwen** [comment]
Ooops, I guess that's a no :-D. Just put a reminder in your calendar 3 months from now to make a PR to take the old parameter out.

**Pull request merged**